test/boost/disk_space_monitor_test.cc

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include <seastar/util/closeable.hh>
+#include <seastar/util/later.hh>
+
+#include "test/lib/log.hh"
+#include "test/lib/scylla_test_case.hh"
+#include "test/lib/tmpdir.hh"
+
+#include "utils/disk_space_monitor.hh"
+
+SEASTAR_THREAD_TEST_CASE(test_disk_space_monitor) {
+    tmpdir dir;
+    utils::disk_space_monitor dsm;
+
+    auto dsm_config = utils::disk_space_monitor::config{
+        .normal_polling_interval = 1ms,
+        .high_polling_interval = 1ms,
+    };
+    dsm.start(dir.path().native(), dsm_config);
+    auto stop_dsm = deferred_stop(dsm);
+
+    int called = 0;
+    auto sub = dsm.subscribe([&] (const utils::disk_space_monitor& dsm) -> future<> {
+        ++called;
+        testlog.info("total_space={} free_space={} utilization={:.1f}%",
+            dsm.total_space(), dsm.free_space(), dsm.disk_utilization() * 100);
+        return make_ready_future();
+    });
+
+    while (!called) {
+        yield().get();
+    }
```

Comment:
What is being tested here?

Reply:
I wonder if this test ever finishes without external action. We could maybe set up a small fuse file system (similar to how we set up a env for s3 tests) to this test, and the test can allocate until monitor calls back the subscriber that will deallocate space, then we can check that the monitor will not call back in the next interval

Reply:
> What is being tested here?

That the disk_space_monitor setup and teardown work correclty, plus the registered callback is being called.

Reply:
> I wonder if this test ever finishes without external action. We could maybe set up a small fuse file system (similar to how we set up a env for s3 tests) to this test, and the test can allocate until monitor calls back the subscriber that will deallocate space, then we can check that the monitor will not call back in the next interval

The callbacks are called periodically and unconditionally.
Why complicate the test?

Reply:
> > I wonder if this test ever finishes without external action. We could maybe set up a small fuse file system (similar to how we set up a env for s3 tests) to this test, and the test can allocate until monitor calls back the subscriber that will deallocate space, then we can check that the monitor will not call back in the next interval
> 
> The callbacks are called periodically and unconditionally.
> Why complicate the test?

Ah usually this kind of interface only calls its subscribers when some event happens. Not saying it's necessarily a bad thing just that I was surprised

Reply:
That was my initial intent but eventually I came to conclusion that inplementing a  generic threshold based event system would be too complicated and it would be more straightforward for each subscriber to manage its own state, if needed, or just implement a stateless algorithm, for example, the maintenance shares controller, or the write guardrail (say to issue warnings above a soft threshold and reject writes above the hard threshold)

Reply:
Next version tests space monitoring (via simulation, not of a real filesystem)